quantization/src/encoded_vectors_pq.rs

```diff
@@ -183,6 +207,62 @@ impl<TStorage: EncodedStorage> EncodedVectorsPQ<TStorage> {
         }
     }
 
+    #[allow(unused)]
+    fn encode_storage_rayon<'a, SB>(
+        data: impl ParallelIterator<Item = &'a [f32]>,
+        storage_builder: Arc<Mutex<&mut SB>>,
+        _vector_parameters: &VectorParameters,
+        vector_division: &[Range<usize>],
+        centroids: &[Vec<f32>],
+        max_kmeans_threads: usize,
+    ) where
+        SB: EncodedStorageBuilder<TStorage> + Send,
+    {
+        let pool = rayon::ThreadPoolBuilder::new()
+            .thread_name(|idx| format!("pq-{idx}"))
+            .num_threads(max_kmeans_threads)
+            .build()
+            .map_err(|e| EncodingError {
+                description: format!("Failed PQ encoding while thread pool init: {e}"),
+            })
+            .unwrap();
+
+        // TODO: minimize allocation, use buffer types
+
+        pool.install(|| {
+            data.map(|vector_data| {
+                let mut encoded_vector = vec![];
+                for range in vector_division.iter() {
+                    let subvector_data = &vector_data[range.clone()];
+                    let mut min_distance = f32::MAX;
+                    let mut min_centroid_index = 0;
+                    for (centroid_index, centroid) in centroids.iter().enumerate() {
+                        let centroid_data = &centroid[range.clone()];
+                        let distance = subvector_data
+                            .iter()
+                            .zip(centroid_data.iter())
+                            .map(|(a, b)| (a - b).powi(2))
+                            .sum::<f32>();
+                        if distance < min_distance {
+                            min_distance = distance;
+                            min_centroid_index = centroid_index;
+                        }
+                    }
+                    encoded_vector.push(min_centroid_index as u8);
+                }
+                encoded_vector
+            })
+            // TODO: must be ordered, check this
+            .for_each_with(storage_builder, |storage_builder, encoded_vector| {
```

Comment:
We must ensure this is ordered.

This breaks with `.par_bridge()` which is currently used in this PoC.